iroha_network/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+}
+
+pub const BUFFER_SIZE: usize = 2048;
+
+pub struct Network {
+    server_connection: TcpStream,
+}
+
+impl Network {
+    /// Creates a new client and connects it to the server.
+    pub fn connect(server_url: &str) -> Result<Network, Box<dyn Error>> {
+        Ok(Network {
+            server_connection: TcpStream::connect(server_url)?,
+        })
+    }
+
+    /// Sends `request` over established and persistent connection and gets `Response`.
+    pub fn send_request(&mut self, request: Request) -> Result<Response, Box<dyn Error>> {
+        Self::send_over_stream(&mut self.server_connection, request)
+    }
+
+    /// Establishes connection, sends `request` closes connection and returns `Response`.
+    pub fn send_request_to(server_url: &str, request: Request) -> Result<Response, Box<dyn Error>> {
+        let mut stream = TcpStream::connect(server_url)?;
+        Self::send_over_stream(&mut stream, request)
+    }
+
+    /// Listens on localhost on port `port`.
+    /// When there is an incoming connections passes it's `Stream` to `handler`.
+    pub fn listen<H>(port: u16, mut handler: H) -> Result<(), Box<dyn Error>>
+    where
+        H: FnMut(Box<dyn Stream>) -> Result<(), Box<dyn Error>>,
+    {
+        let listener = TcpListener::bind(SocketAddr::new(IpAddr::from_str("127.0.0.1")?, port))?;
+        while let Some(stream) = listener.incoming().next() {
+            handler(Box::new(stream?))?;
+        }
+        Ok(())
+    }
+
+    /// Helper function to call inside `listen` `handler` function to parse and send response.
+    /// The `handler` specified here will need to generate `Response` from `Request`.
+    pub fn handle_message<H>(mut stream: impl Stream, mut handler: H) -> Result<(), Box<dyn Error>>
+    where
+        H: FnMut(Request) -> Result<Response, Box<dyn Error>>,
+    {
+        let mut buffer = [0u8; BUFFER_SIZE];
+        let read_size = stream.read(&mut buffer).expect("Request read failed.");
+        let response = handler(
+            buffer[..read_size]
+                .to_vec()
+                .try_into()
+                .map_err(|_| "Failed to parse message.")?,
+        )?;
+        stream.write_all(&response)?;
+        stream.flush()?;
+        Ok(())
+    }
+
+    /// Function for internal use to send messages over TcpStream.
+    fn send_over_stream(
+        stream: &mut TcpStream,
+        request: Request,
+    ) -> Result<Response, Box<dyn Error>> {
+        let payload: Vec<u8> = request.into();
+        stream.write_all(&payload)?;
+        stream.flush()?;
+        let mut buffer = [0u8; BUFFER_SIZE];
+        let read_size = stream.read(&mut buffer)?;
+        Ok(buffer[..read_size].to_vec())
+    }
+}
+
+pub trait Stream: Read + Write + Send {}
+impl<T> Stream for T where T: Read + Write + Send {}
+
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Request {
+    url: String,
+    payload: Vec<u8>,
+}
+
+impl Request {
+    pub fn new(url: String, payload: Vec<u8>) -> Request {
```

Comment:
Maybe we should use URI naming https://en.wikipedia.org/wiki/Uniform_Resource_Identifier and add documentation about the format?